### PR TITLE
Feature/auth backend

### LINKED ISSUE #20: Authentication with salt + hash
Authenticate users with a username and password. The password must be salted and hashed before being stored in the database.

### PR DESCRIPTION
**Pull request by jake-shasteen**
Resolves #20 

I tested this with Postman -- if you haven't used it before, it's a great way to test APIs. The three routes that I've implemented should work:
/api/signup
/api/login
/api/username

In order to test you'll have to have a local copy of mysql or mariadb running on port 3306, with a user called 'root' with a blank password, and a database called 'geekmapper'.

You can change the SQL username, password, database name, and port with environment variables, but they all have default values as above. Those environment variables are: SQLPASS, SQLUSER, SQLPORT, SQLDBNAME.

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
+'use strict';
+
```
Comment (Malkleth) [inline server/controllers/userController.js]: look at me!

File: server/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        res.status( 500 ).send( 'Error creating new user' );
+        return null;
+      // We found the user -- it already exists
+      } else if ( user === 'exists' ) {
+        return null;
+      }
+      // The user didn't exist, and we were able to create a new one.
+      // create token to send back for auth
+      var token = helpers.encode( user );
+      res.json({ token });
```
Comment (jake-shasteen) [inline server/controllers/userController.js]: Send status of 201: Created

File: server/models/users.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      type: DataTypes.STRING,
+      unique: true,
+    },
+    password: DataTypes.STRING,
+    email: {
+      type: DataTypes.STRING,
+      validate: {
+        isEmail: true,
+      },
+    },
```
Comment (jake-shasteen) [inline server/models/users.js]: add `accountType` for permission levels as TINY INT

File: server/lib/helpers.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  },
+  
+  errorHandler: function ( error, req, res, next ) {
+    // send error message to client
+    // message for graceful error handling on app
+    res.status( 500 ).send({ error: error.message });
+  },
+
+  encode: function( user ) {
+    return jwt.encode( user, process.env.JWT_SECRET );
```
Comment (jake-shasteen) [inline server/lib/helpers.js]: redact `password` property of `user`

File: server/controllers/userController.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    const password = req.body.password;
+    User.findOne( { where: { username } } )
+    .then( ( user ) => {
+      if( user ) {
+        user.validPassword( password )
+        .then( ( match ) => {
+          if( match ) {
+            // Right password
+            const token = helpers.encode( user );
+            return res.json({ token });
```
Comment (jake-shasteen) [inline server/controllers/userController.js]: Send 201 status code

**Comment by jake-shasteen** [comment]
This should be good to go now. Fixed the things pointed out in our walkthrough yesterday.

**Pull request merged**